File: php-plugin/src/main/java/fr/cnumr/php/checks/AvoidUsingGlobalVariablesCheck.java
package fr.cnumr.php.checks;

import org.sonar.check.Priority;
import org.sonar.check.Rule;
import org.sonar.plugins.php.api.tree.declaration.FunctionDeclarationTree;
import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;

import java.util.regex.Pattern;

@Rule(
        key = "D4",
        name = "Developpement",
        description = AvoidUsingGlobalVariablesCheck.ERROR_MESSAGE,
        priority = Priority.MINOR,
        tags = {"bug"})

public class AvoidUsingGlobalVariablesCheck extends PHPVisitorCheck {

    public static final String ERROR_MESSAGE = "Prefer local variables to globals";
    public static final String GLOBALS_PATTERN = "^.*(global \\$|\\$GLOBALS).*$";
    private Pattern pattern;
    String format = GLOBALS_PATTERN;
    @Override
    public void init() {
        pattern = Pattern.compile(format,Pattern.CASE_INSENSITIVE);
    }
    @Override
    public void visitFunctionDeclaration(FunctionDeclarationTree tree) {
        if(pattern.matcher(tree.body().toString()).matches()){
            context().newIssue(this, tree, String.format(ERROR_MESSAGE, tree.body().toString()));
        }
    }
}

File: php-plugin/src/test/resources/checks/AvoidUsingGlobalVariablesCheck.php
<?php
$a = 1;
$b = 2;
function somme() { // NOK {{Prefer local variables to globals}}
    $GLOBALS['b'] = $GLOBALS['a'] + $GLOBALS['b'];
}
somme();
echo $b;

function somme2() { // NOK {{Prefer local variables to globals}}
    global $a, $b;
    $b = $a + $b;
}
somme2();
echo $b;
?>
